### PR TITLE
Implement ellipsizing of text in the engine

### PR DESCRIPTION
**Pull request by jason-simmons**
See https://github.com/flutter/flutter/issues/4478

### CODE REVIEW

File: sky/engine/core/rendering/line/BreakingContextInlineHeaders.h [hunk 1]
```diff
@@ -409,6 +409,14 @@ inline float measureHyphenWidth(RenderText* renderer, const Font& font, TextDire
         style->hyphenString().string(), style, style->direction()));
 }
 
+inline float measureEllipsisWidth(RenderText* renderer, const Font& font)
+{
+    const UChar ellipsis = 0x2026;
```
Comment (abarth) [inline sky/engine/core/rendering/line/BreakingContextInlineHeaders.h]: There's already a constant for this in CharacterNames.h: horizontalEllipsis

File: sky/engine/core/rendering/line/BreakingContextInlineHeaders.h [hunk 2]
```diff
@@ -537,6 +561,7 @@ inline bool BreakingContext::handleText(WordMeasurements& wordMeasurements, bool
 
             wordMeasurement.width = additionalTempWidth + wordSpacingForWordMeasurement;
             additionalTempWidth += lastSpaceWordSpacing;
+
```
Comment (abarth) [inline sky/engine/core/rendering/line/BreakingContextInlineHeaders.h]: spurious change

File: sky/engine/core/rendering/line/BreakingContextInlineHeaders.h [hunk 3]
```diff
@@ -584,6 +609,13 @@ inline bool BreakingContext::handleText(WordMeasurements& wordMeasurements, bool
                             wordMeasurement.width = charWidth;
                         }
                     }
+
+                    if (ellipsizeMode) {
+                      // Break the line at the position where an ellipsis would fit.
```
Comment (abarth) [inline sky/engine/core/rendering/line/BreakingContextInlineHeaders.h]: four space indent

File: sky/engine/core/rendering/style/RenderStyle.h [hunk 4]
```diff
@@ -627,6 +627,9 @@ class RenderStyle: public RefCounted<RenderStyle> {
     // Apple-specific property getter methods
     EPointerEvents pointerEvents() const { return static_cast<EPointerEvents>(inherited_flags._pointerEvents); }
 
+    // Flutter extension property getter methods
+    EFlutterLineBreakMode flutterLineBreakMode() const { return static_cast<EFlutterLineBreakMode>(rareInheritedData->m_flutterLineBreakMode); }
```
Comment (abarth) [inline sky/engine/core/rendering/style/RenderStyle.h]: I'd remove the "flutter" from these names.  We own this code now.  (For better or worse)

File: sky/engine/core/rendering/style/RenderStyleConstants.h [hunk 5]
```diff
@@ -154,6 +154,10 @@ enum ETextAlign {
     LEFT, RIGHT, CENTER, JUSTIFY, TASTART, TAEND,
 };
 
+enum EFlutterLineBreakMode {
+    STANDARD, ELLIPSIZE
```
Comment (abarth) [inline sky/engine/core/rendering/style/RenderStyleConstants.h]: C++ enums are dumb.  This creates a global symbol called STANDARD.  In CSS, this is called "clip"...  Dunno.

File: lib/ui/text.dart [hunk 6]
```diff
@@ -423,12 +439,14 @@ class ParagraphStyle {
     TextAlign textAlign,
     FontWeight fontWeight,
     FontStyle fontStyle,
+    TextLineBreakMode textLineBreakMode,
```
Comment (abarth) [inline lib/ui/text.dart]: We generally try to use names inspired by CSS...  Maybe call this "textOverflow"

File: lib/ui/text.dart [hunk 7]
```diff
@@ -93,6 +93,15 @@ enum TextAlign {
   center
 }
 
+/// How to break lines in a paragraph
+enum TextLineBreakMode {
```
Comment (abarth) [inline lib/ui/text.dart]: TextOverflow

File: lib/ui/text.dart [hunk 8]
```diff
@@ -93,6 +93,15 @@ enum TextAlign {
...
+  /// Default line break mode
+  standard,
```
Comment (abarth) [inline lib/ui/text.dart]: clip

File: lib/ui/text.dart [hunk 9]
```diff
@@ -93,6 +93,15 @@ enum TextAlign {
...
+
+  /// Render as a single line that is ellipsized if it is too wide to fit
+  ellipsize
```
Comment (abarth) [inline lib/ui/text.dart]: ellipsis

See https://developer.mozilla.org/en-US/docs/Web/CSS/text-overflow

File: lib/ui/text.dart [hunk 10]
```diff
@@ -423,12 +439,14 @@ class ParagraphStyle {
     TextAlign textAlign,
     FontWeight fontWeight,
     FontStyle fontStyle,
+    TextLineBreakMode textLineBreakMode,
     String fontFamily,
     double fontSize,
     double lineHeight
```
Comment (abarth) [inline lib/ui/text.dart]: Eventually, we'll want to add a "lineCount" field here that says how many lines of text to show.  We don't need to add that in this patch (obviously), but that's what's guiding some of these naming choices.

**Comment by jason-simmons** [comment]
Removed FlutterLineBreakMode and reused the CSS text-overflow property.

PTAL

File: sky/engine/core/rendering/InlineTextBox.cpp [hunk 11]
```diff
@@ -450,9 +450,19 @@ void InlineTextBox::paint(PaintInfo& paintInfo, const LayoutPoint& paintOffset,
         string.narrow(m_start, length);
     maximumLength = renderer().textLength() - m_start;
 
+    StringBuilder charactersWithEllipsis;
+    if (hasAddedEllipsis()) {
+      const UChar ellipsis = 0x2026;
```
Comment (abarth) [inline sky/engine/core/rendering/InlineTextBox.cpp]: Did you want to use the named constant here as well?

Comment (jason-simmons) [inline sky/engine/core/rendering/InlineTextBox.cpp]: done

**Comment by Hixie** [comment]
I'd rather the ellipsis was a configurable string, but that can wait til later I guess.

**Pull request merged**